### ISSUE TITLE
ingress-sds issues/possible memory-leak

### ISSUE THREAD

**Issue by jcetkov** [istio/istio#13197]
**Describe the bug**
I observed several possibly related issues while using the ingress-sds
1) on larger cluster the ingress-sds seems to leak memory, in magnitude of 2-3 GBs over the span of 3 or 4 days; this seems to correlated with bursts of messages
```
2019-04-09T23:39:34.206560Z	info	SDS: push key/cert pair from node agent to proxy: "router~10.24.24.97~istio-ingressgateway-598b5d6fdc-7qrgj.istio-system~istio-system.svc.cluster.local-39832"
```
10s in a second, thousands a day
```
k logs -n istio-system istio-ingressgateway-598b5d6fdc-7qrgj  ingress-sds --tail 10000 | grep "push key/cert pair from node agent to proxy" | grep "2019-04-09T23:39:34" | wc -l
      38
k logs -n istio-system istio-ingressgateway-598b5d6fdc-7qrgj  ingress-sds  | grep "push key/cert pair from node agent to proxy" | grep "2019-04-09" | wc -l
   21975
```

2) I was able to reproduce similar behavior locally, if I create secret with an empty cert. In such case, the loop never ends, and the process quickly drains memory, until the secret is removed.
(The leak from above can be to some extent reproduced by having this running for a time period, however fixing the cert results in most of the memory released after period of time) 
```
k logs -n istio-system istio-ingressgateway-6796fc8477-d48qv  ingress-sds | grep "2019-04-09T23:52:19" | grep "push key/cert pair from node agent to proxy" | wc -l
     554
```

3) In at least one case, on removal of the secret in question, the sds agent died 
```
 panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x11d88db]

goroutine 631 [running]:
istio.io/istio/security/pkg/nodeagent/sds.pushSDS(0xc4297c4200, 0x15db980, 0x0)
    /workspace/go/src/istio.io/istio/security/pkg/nodeagent/sds/sdsservice.go:344 +0x1db
istio.io/istio/security/pkg/nodeagent/sds.(*sdsservice).StreamSecrets(0xc4203f7040, 0x168ab00, 0xc421028100, 0x0, 0x0)
    /workspace/go/src/istio.io/istio/security/pkg/nodeagent/sds/sdsservice.go:213 +0x7aa
istio.io/istio/vendor/github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2._SecretDiscoveryService_StreamSecrets_Handler(0x13b9180, 0xc4203f7040, 0x1686480, 0xc420224580, 0x1f55830, 0xc4233ee300)
    /workspace/go/src/istio.io/istio/vendor/github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2/sds.pb.go:175 +0xb2
istio.io/istio/vendor/google.golang.org/grpc.(*Server).processStreamingRPC(0xc420311680, 0x168c7e0, 0xc42065a900, 0xc4370be000, 0xc420406720, 0x1f1f7e0, 0x0, 0x0, 0x0)
    /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:1124 +0x911
istio.io/istio/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420311680, 0x168c7e0, 0xc42065a900, 0xc4370be000, 0x0)
    /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:1212 +0x12b1
istio.io/istio/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4200445c0, 0xc420311680, 0x168c7e0, 0xc42065a900, 0xc4370be000)
    /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:686 +0x9f
created by istio.io/istio/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
    /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:684 +0xa1
```

**Expected behavior**
The sds should not leak memory/fail on secret with missing certificate.   

**Steps to reproduce the bug**
follow https://istio.io/docs/tasks/traffic-management/secure-ingress/sds/
when creating the secret, pass an empty file for the cert

```
touch empty.cert.pem
kubectl create -n istio-system secret generic httpbin-credential \
--from-file=key=httpbin.example.com/3_application/private/httpbin.example.com.key.pem \
--from-file=cert=empty.cert.pem
```

**Version**
```
bin/istioctl version --remote
client version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.1"}
citadel version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}
galley version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}
ingressgateway version: version.BuildInfo{Version:"bdda7cfcf5ba1397e6e0e2629d53114c9ea8fc14", GitRevision:"bdda7cfcf5ba1397e6e0e2629d53114c9ea8fc14", User:"mjog", Host:"devinstance.c.mixologist-142215.internal", GolangVersion:"go1.10.1", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.0-snapshot.4-592-gbdda7cf"}
pilot version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}
policy version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}
sidecar-injector version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}
telemetry version: version.BuildInfo{Version:"1.1.2", GitRevision:"2b1331886076df103179e3da5dc9077fed59c989-dirty", User:"root", Host:"35adf5bb-5570-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.1.1"}

kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Installation**
```
helm template install/kubernetes/helm/istio --set global.mtls.enabled=true --set gateways.istio-egressgateway.enabled=false --set gateways.istio-ingressgateway.sds.enabled=true --name istio --namespace istio-system > ./istio.yaml
kubectl apply -f istio.yaml
```

**Environment**
docker edge for mac (local)
custom build cluster over azure nodes (remote, issue 1))

**Cluster state**
empty cluster with freshly installed istio (for the local reproduction)
cluster served by CI pipeline with handful of random applications (remote)

**Comment by jcetkov** [comment]
Looks like I pinpointed the root cause of the issue 1). It seems to be gateway referencing `credentialName` which doesn't exist.

To confirm locally, I used a script to create 500 number copies of the following config 
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: fortiobin1-gateway
  namespace: tom-namespace
spec:
  selector:
    istio: ingressgateway # use Istio default gateway implementation
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      credentialName: fortiobin1-secret # must be the same as secret
    hosts:
    - fortiobin1.example.com
```
(no secrets or any other related CRDs were created)
This is a graph of memory usage of the sds container since that moment

<img width="1310" alt="Screen Shot 2019-04-10 at 3 18 17 PM" src="https://user-images.githubusercontent.com/6686220/55917472-46142400-5ba4-11e9-98ee-067b7b288068.png">

**Comment by JimmyCYJ** [comment]
Thanks for the investigation! The ingress gateway agent does not follow SDS protocol strictly when credential does not exist. The ingress gateway agent should not send update to every SDS request with empty cert. I will fix this.

**Comment by jcetkov** [comment]
it also seems to loop and draw memory in case the secret doesn't exist, what we observed in the large cluster was the memory function looked liked in the picture above, only over longer span of time (essentially hogging memory until killed or restarted) while continuously, although not as rapidly logging
```
2019-04-10T22:28:01.772406Z	info	SDS: connection with "router~10.1.4.62~istio-ingressgateway-6796fc8477-d48qv.istio-system~istio-system.svc.cluster.local-96940" terminated rpc error: code = Canceled desc = context canceled
2019-04-10T22:28:01.969799Z	error	Failed to generate secret for proxy "router~10.1.4.62~istio-ingressgateway-6796fc8477-d48qv.istio-system~istio-system.svc.cluster.local-96941": cannot find secret fortiobin490-secret for ingress gateway
2019-04-10T22:28:01.969960Z	error	Failed to get secret for proxy "router~10.1.4.62~istio-ingressgateway-6796fc8477-d48qv.istio-system~istio-system.svc.cluster.local" connection "router~10.1.4.62~istio-ingressgateway-6796fc8477-d48qv.istio-system~istio-system.svc.cluster.local-96941" from secret cache: cannot find secret fortiobin490-secret for ingress gateway
```

**Comment by duderino** [comment]
Sorry for the reassignment spam.  The github UI kinda freaked out on me.  Wasn't intentional.

@JimmyCYJ if you can get a PR out by 5pm PT tomorrow April 11th, we can get it into 1.1.3.  Otherwise it'll have to go out for 1.1.4 about a week later

**Comment by JimmyCYJ** [comment]
> @JimmyCYJ if you can get a PR out by 5pm PT tomorrow April 11th, we can get it into 1.1.3. Otherwise it'll have to go out for 1.1.4 about a week later

Thanks for the info, I will try to get a PR out by 5pm April 11th.

**Comment by JimmyCYJ** [comment]
I am able to reproduce this memory leak issue by creating 70 TLS gateway without credentials.

**Comment by JimmyCYJ** [comment]
Confirmed with @jcetkov that the problem is fixed. I am going to close this issue.

**Comment by duderino** [comment]
The fix made it into 1.1.3
